server/src/test/java/org/elasticsearch/action/admin/cluster/desiredbalance/TransportGetDesiredBalanceActionIntegrationTests.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.action.admin.cluster.desiredbalance;
+
+import org.elasticsearch.action.bulk.BulkRequestBuilder;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.test.ESIntegTestCase;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class TransportGetDesiredBalanceActionIntegrationTests extends ESIntegTestCase {
+
+    public void testDesiredBalanceOnMultiNodeCluster() throws Exception {
+        internalCluster().startMasterOnlyNode();
+        internalCluster().startDataOnlyNodes(randomIntBetween(2, 5));
+
+        String index = "test";
+        int numberOfShards = 2;
+        int numberOfReplicas = 1;
+        createIndex(index, Settings.builder().put("index.number_of_shards", 2).put("index.number_of_replicas", 1).build());
+
+        indexData(index);
+
+        // Wait until all replicas have started
+        waitUntil(() -> {
+            IndexRoutingTable indexRoutingTable = client().admin().cluster().prepareState().get().getState().routingTable().index(index);
+            for (int i = 0; i < indexRoutingTable.size(); i++) {
+                IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(i);
+                for (int j = 0; j < shardRoutingTable.size(); j++) {
+                    ShardRouting shard = shardRoutingTable.shard(j);
+                    if (shard.state() != ShardRoutingState.STARTED) {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        });
```

Comment:
This should be enforced by a `createIndex`.
Was it not the case for some of the test runs?

Reply:
Uh, my bad, we only wait for primaries. Create with `waitForActiveShards=ALL` will do this

Reply:
Or else use the health API to wait for a green status.

Reply:
No, but I was able to achieve this was David's advice of using the cluster health API :+1: